include/circt/Dialect/MSFT/MSFT.td

```diff
@@ -33,4 +33,29 @@ def MSFTDialect : Dialect {
 class MSFTOp<string mnemonic, list<OpTrait> traits = []> :
     Op<MSFTDialect, mnemonic, traits>;
 
+def DeviceType : I32EnumAttr<"DeviceType",
+  "Type of device at physical location", [
+    I32EnumAttrCase<"M20K", 1>,
+    I32EnumAttrCase<"DSP", 2>,
+    I32EnumAttrCase<"EC", 3>,
+  ]>;
+
+class MSFT_Attr<string name, string baseCppClass = "::mlir::Attribute">
+    : AttrDef<MSFTDialect, name, baseCppClass> {
+  let mnemonic = ?;
+}
+
+def PhysLocation : MSFT_Attr<"PhysLocation"> {
+  let summary = "Descibes a physical location on a device";
+  let description = [{
+    Annotate a particular entity within an op with the location of the device
+    on an FPGA to which it should mapped.
+  }];
+  let mnemonic = "physloc";
+  let parameters = (ins
+    "DeviceTypeAttr":$devType,
+    "uint64_t":$x, "uint64_t":$y, "uint64_t":$num,
+    "::llvm::StringRef":$entity);
```

Comment:
Drive by, but what is the lifetime of `entity` here? Should this be using `StringRefParameter` instead?

Reply:
D'oh! I was making an assumption which isn't valid. Good catch -- thanks!

Reply:
Ended up going a different way with this which obviated the need for this parameter.